script.module.kodiswift/lib/kodiswift/xbmcmixin.py

```diff
@@ -0,0 +1,532 @@
+        """Displays the keyboard input window to the user. If the user does not
+        cancel the modal, the value entered by the user will be returned.
+
+        :param default: The placeholder text used to prepopulate the input
+                        field.
+        :param heading: The heading for the window. Defaults to the current
+                        addon's name. If you require a blank heading, pass an
+                        empty string.
+        :param hidden: Whether or not the input field should be masked with
+                       stars, e.g. a password field.
+        """
+        if heading is None:
+            heading = self.addon.getAddonInfo('name')
+        if default is None:
+            default = ''
+        keyboard = xbmc.Keyboard(default, heading, hidden)
+        keyboard.doModal()
+        if keyboard.isConfirmed():
+            return keyboard.getText()
+
+    def notify(self, msg='', title=None, delay=5000, image=''):
+        """Displays a temporary notification message to the user. If
+        title is not provided, the plugin name will be used. To have a
+        blank title, pass '' for the title argument. The delay argument
+        is in milliseconds.
+        """
+        if not msg:
+            log.warning('Empty message for notification dialog')
+        if title is None:
+            title = self.addon.getAddonInfo('name')
+        xbmc.executebuiltin('Kodi.Notification("%s", "%s", "%s", "%s")' %
+                            (msg, title, delay, image))
+
+    def _listitemify(self, item):
+        """Creates an kodiswift.ListItem if the provided value for item is a
+        dict. If item is already a valid kodiswift.ListItem, the item is
+        returned unmodified.
+        """
+        info_type = self.info_type if hasattr(self, 'info_type') else 'video'
+
+        # Create ListItems for anything that is not already an instance of
+        # ListItem
+        if not hasattr(item, 'as_tuple'):
+            if 'info_type' not in item:
+                item['info_type'] = info_type
+            item = kodiswift.ListItem.from_dict(**item)
+        return item
+
+    @staticmethod
+    def _add_subtitles(subtitles):
+        """Adds subtitles to playing video.
+
+        Warnings:
+            You must start playing a video before calling this method or it
+            will raise and Exception after 30 seconds.
+
+        Args:
+            subtitles (str): A URL to a remote subtitles file or a local
+                filename for a subtitles file.
+        """
+        # This method is named with an underscore to suggest that callers pass
+        # the subtitles argument to set_resolved_url instead of calling this
+        # method directly. This is to ensure a video is played before calling
+        # this method.
+        player = xbmc.Player()
+        for _ in range(30):
+            if player.isPlaying():
+                break
+            time.sleep(1)
```

Comment:
please check if an abort is requested, see http://kodi.wiki/view/Service%20addons